OneCommunityGlobal/HGNRest
# Shengwei add cache for the badge APIs

# Description
![image](https://github.com/OneCommunityGlobal/HGNRest/assets/69882989/e70e17fd-ad98-4bda-b86b-43a1c2dcdb54)

## Related PRS (if any):


## Main changes explained:
- Modify file src/controllers/badgeController.js for adding cache handling in all badge CRUD operation.


## How to test:
1. check into the current branch
2. do `npm install` and `...` to run this PR locally
3. Clear site data/cache
4. log as admin/owner user
5. open browser -> right-click on page and click inspect -> go to inspect mode -> go to network
6. go to dashboard→ others -> badge management 
   - verify that any call to `/api/badge` has significant performance improvement after the initial call. 
   - You could refresh the page and check the Time in the Network tab under the Inspect mode. (Should have around 900ms end-to-end response time in the first call and around 90ms in any later calls) 
Initial API call
![image](https://github.com/OneCommunityGlobal/HGNRest/assets/69882989/e4d183dc-0fb9-489f-802b-5760e7df5fd0)
API call with cache
![image](https://github.com/OneCommunityGlobal/HGNRest/assets/69882989/5265aaf3-a830-4747-b5a0-4f0a050a89e9)

7. go to badge management -> click on badge development -> **CREATE** new badge -> verify badge API should remove cache 
  - Clear the network log for better visibility 
  
![image](https://github.com/OneCommunityGlobal/HGNRest/assets/69882989/8b36580b-c555-4f0d-bd2c-44419038f941)

  - You should create a new badge and see two calls to `/api/badge`
  - You should read the performance on call that used the GET method. The performance should be significantly slower than cached calls since we remove the cache after the badge update.
  - Refresh the page and verify any later calls to  `/api/badge` have significantly faster performance.
  
![image](https://github.com/OneCommunityGlobal/HGNRest/assets/69882989/99638dff-114b-41a6-8899-f723faa7e5da)
![image](https://github.com/OneCommunityGlobal/HGNRest/assets/69882989/a71b06ec-41c3-4461-86ec-9e8b67412a41)

8. go to badge management -> click on badge development -> **DELETE** newly created badge -> verify badge API should remove cache 
  - You should delete the newly created badge and see one call to `/api/badge` with the GET method
  - You should read the performance on call that used the GET method. The performance should be significantly slower than cached calls since we remove the cache after the badge update.
  ![image](https://github.com/OneCommunityGlobal/HGNRest/assets/69882989/5c175f7e-c7a2-4110-8be4-bdb66deeaf82)
  - Refresh the page and verify any later calls to  `/api/badge` have significantly faster performance.


## Screenshots or videos of changes:
No Sound

https://github.com/OneCommunityGlobal/HGNRest/assets/69882989/88bbcc6d-adc7-43f5-b412-a547b4fa200e


## Note:
Include the information the reviewers need to know.
